### ISSUE TITLE
Non-CONFIG find_package() is very broken in vcpkg

### ISSUE THREAD

**Issue by ned14** [microsoft/vcpkg#23460]
**Describe the bug**

vcpkg's Module based `find_package()` causes the `FindXXX.cmake` in the cmake to get used, and it sometimes will choose the system installed edition of the package rather than the vcpkg installed edition. I particularly ran into this problem when combining:

1. A custom triplet to force static libraries on Windows (why this isn't the default on Windows to match POSIX is beyond me!).
2. `find_package(OpenSSL)` or `find_package(ZLIB)` or `find_package(BZip2)`, which are all non-CONFIG and thus cause fallback onto the cmake's `FindXXX.cmake`.
3. The system installed version of that library is very old i.e. binary incompatible.

**Environment**
All platforms I tested (Windows, Linux, MSVC, GCC, clang).

**To Reproduce**
See above. I was using vcpkg in manifest mode and with the vcpkg toolchain file injected.

**Expected behavior**
I never want system installed libraries to ever be chosen by `find_package()`, not ever. Only the vcpkg ones.

**Additional context**
I hacked around the problem using this evil function:

```
set(VCPKG_ROOT_DIR "${CMAKE_BINARY_DIR}/vcpkg" CACHE INTERNAL "vcpkg root path" FORCE)
set(VCPKG_INSTALLED_DIR "${CMAKE_BINARY_DIR}/vcpkg_installed" CACHE INTERNAL "vcpkg installation path" FORCE)
set(VCPKG_TARGET_TRIPLET "x64-mdx-config" CACHE INTERNAL "vcpkg target triplet" FORCE)
set(VCPKG_MANIFEST_MODE On CACHE INTERNAL "vcpkg manifest mode" FORCE)
set(VCPKG_MANIFEST_DIR "${CMAKE_SOURCE_DIR}" CACHE INTERNAL "vcpkg manifest dir" FORCE)
function(vcpkg_find_package pkg)
  cmake_parse_arguments(VFP "REQUIRED" "VCPACKAGE_NAME" "TARGETS;LIBRARY_NAMES;WIN_LIBRARY_NAMES" ${ARGN})
  set(VFP_UNPARSED_ARGUMENTS2 ${VFP_UNPARSED_ARGUMENTS})
  if(NOT "QUIET" IN_LIST VFP_UNPARSED_ARGUMENTS2)
    list(APPEND VFP_UNPARSED_ARGUMENTS2 "QUIET")
  endif()
  if(${pkg}_FOUND)
    return()
  endif()
  if(WIN32 AND DEFINED VFP_WIN_LIBRARY_NAMES)
    set(VFP_LIBRARY_NAMES ${VFP_WIN_LIBRARY_NAMES})
  endif()
  # vcpkg sets CMAKE_PREFIX_PATH with its installed packages, so disable finding packages from anywhere else
  find_package(${pkg} ${VFP_UNPARSED_ARGUMENTS2}
    NO_PACKAGE_ROOT_PATH
    # NO_CMAKE_PATH
    NO_CMAKE_ENVIRONMENT_PATH
    NO_SYSTEM_ENVIRONMENT_PATH
    NO_CMAKE_PACKAGE_REGISTRY
    NO_CMAKE_SYSTEM_PATH
    NO_CMAKE_SYSTEM_PACKAGE_REGISTRY
    HINTS "${CMAKE_BINARY_DIR}/vcpkg_installed/${VCPKG_TARGET_TRIPLET}"
  )
  if(NOT ${pkg}_FOUND)
    if(NOT "CONFIG" IN_LIST VFP_UNPARSED_ARGUMENTS2)
      # It would seem that there is a bug in vcpkg when you use a custom triplet, which we
      # are forced to do to solve the problem of generating statically linked libraries
      # in a consistent way across packages, which causes the non-generation of the shim
      # cmake scripting for find_package() to work if the package is not a native CONFIG
      # package. If we reach here, this is such a package, so use find_library() to
      # reconstruct the imported cmake targets which would ordinarily have been shimmed by
      # vcpkg
      #message("   ${pkg}_ROOT=${${pkg}_ROOT}\n   CMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH}\n   CMAKE_FRAMEWORK_PATH=${CMAKE_FRAMEWORK_PATH}\n   CMAKE_APPBUNDLE_PATH=${CMAKE_APPBUNDLE_PATH}\n   ${pkg}_DIR=${${pkg}_DIR}\n   PATH=$ENV{PATH}\n   CMAKE_SYSTEM_PREFIX_PATH=${CMAKE_SYSTEM_PREFIX_PATH}\n   CMAKE_SYSTEM_FRAMEWORK_PATH=${CMAKE_SYSTEM_FRAMEWORK_PATH}\n   CMAKE_SYSTEM_APPBUNDLE_PATH=${CMAKE_SYSTEM_APPBUNDLE_PATH}")
      if(NOT DEFINED VFP_LIBRARY_NAMES OR NOT DEFINED VFP_TARGETS)
        message(FATAL_ERROR "FATAL: Package ${pkg} was not found within vcpkg, if there was no indication above of build failure and the problem is in fact a find failure please supply a LIBRARY_NAMES parameter to tell me what library names to search for, and a TARGETS parameter to tell me what imported targets to create, to induce the workaround for find failure.")
      endif()
      list(LENGTH VFP_LIBRARY_NAMES len1)
      list(LENGTH VFP_TARGETS len2)
      if(NOT len1 EQUAL len2)
        message(FATAL_ERROR "FATAL: Package ${pkg} was not found within vcpkg, the LIBRARY_NAMES parameter must be the same length as the TARGETS parameter as each library name is assigned to its imported target.")
      endif()
      set(${pkg}_FOUND TRUE)
      foreach(idx RANGE ${len1})
        if(idx EQUAL ${len1})
          break()
        endif()
        list(GET VFP_LIBRARY_NAMES ${idx} lib)
        list(GET VFP_TARGETS ${idx} target)
        find_library(${lib}_LIB NAMES ${lib}
          NO_PACKAGE_ROOT_PATH
          # NO_CMAKE_PATH
          NO_CMAKE_ENVIRONMENT_PATH
          NO_SYSTEM_ENVIRONMENT_PATH
          NO_CMAKE_SYSTEM_PATH
          HINTS "${CMAKE_BINARY_DIR}/vcpkg_installed/${VCPKG_TARGET_TRIPLET}/lib"
        )
        if(NOT ${lib}_LIB)
          set(${pkg}_FOUND FALSE)
          message(STATUS "NOTE: Failed to reconstruct vcpkg imported target ${target} as ${lib} was not found in the vcpkg installed path. I searched '${CMAKE_BINARY_DIR}/vcpkg_installed/${VCPKG_TARGET_TRIPLET}/lib' and '${CMAKE_PREFIX_PATH}'")
        else()
          if(NOT TARGET ${target})
            add_library(${target} UNKNOWN IMPORTED)
          endif()
          set_target_properties(${target} PROPERTIES
            INTERFACE_INCLUDE_DIRECTORIES "${CMAKE_BINARY_DIR}/vcpkg_installed/${VCPKG_TARGET_TRIPLET}/include"
            IMPORTED_LOCATION "${${lib}_LIB}"
          )
          message(STATUS "NOTE: Reconstructed failed to find_package(${pkg}) in vcpkg imported target ${target} using found library ${${lib}_LIB}")
        endif()
      endforeach()
      if(${pkg}_FOUND)
        set(${pkg}_FOUND TRUE PARENT_SCOPE)
        set(${pkg}_SKIP_VCPKG_INSTALL TRUE CACHE BOOL "")
      endif()
    endif()
    if(NOT ${pkg}_FOUND)
      if(VFP_REQUIRED)
        find_package(${pkg} ${VFP_UNPARSED_ARGUMENTS} REQUIRED
          NO_PACKAGE_ROOT_PATH
          # NO_CMAKE_PATH
          NO_CMAKE_ENVIRONMENT_PATH
          NO_SYSTEM_ENVIRONMENT_PATH
          NO_CMAKE_PACKAGE_REGISTRY
          NO_CMAKE_SYSTEM_PATH
          NO_CMAKE_SYSTEM_PACKAGE_REGISTRY
          HINTS "${CMAKE_BINARY_DIR}/vcpkg_installed/${VCPKG_TARGET_TRIPLET}"
        )
      else()
        find_package(${pkg} ${VFP_UNPARSED_ARGUMENTS}
          NO_PACKAGE_ROOT_PATH
          # NO_CMAKE_PATH
          NO_CMAKE_ENVIRONMENT_PATH
          NO_SYSTEM_ENVIRONMENT_PATH
          NO_CMAKE_PACKAGE_REGISTRY
          NO_CMAKE_SYSTEM_PATH
          NO_CMAKE_SYSTEM_PACKAGE_REGISTRY
          HINTS "${CMAKE_BINARY_DIR}/vcpkg_installed/${VCPKG_TARGET_TRIPLET}"
        )
      endif()
    endif()
  endif()
  if(${pkg}_FOUND AND DEFINED VFP_TARGETS)
    set_target_properties(${VFP_TARGETS} PROPERTIES
      MAP_IMPORTED_CONFIG_MINSIZEREL Release
      MAP_IMPORTED_CONFIG_RELWITHDEBINFO Release
    )
  endif()
endfunction()
```

As examples of use:

```
vcpkg_find_package(OpenSSL REQUIRED VCPACKAGE_NAME openssl TARGETS OpenSSL::SSL OpenSSL::Crypto LIBRARY_NAMES ssl crypto WIN_LIBRARY_NAMES libssl libcrypto)
vcpkg_find_package(ZLIB REQUIRED VCPACKAGE_NAME zlib TARGETS ZLIB::ZLIB LIBRARY_NAMES z WIN_LIBRARY_NAMES zlib)
```

This suggests that the vast majority of non-CONFIG cmake packages could be easily retrofitted to never ever search system installs (some `FindXXX.cmake` scripts offer a custom cmake variable to disable system package searching, but not all do, and the naming of that override variable is not consistent).

Alternatively, all non-CONFIG cmake packages could be required to become CONFIG based, and all this would go away. I would greatly prefer the latter. It could be implemented as a package upgrade rule.

**Comment by dg0yt** [comment]
You seem to use vcpkg in a wrong way.

> A custom triplet to force static libraries on Windows

Why not use one of the existing static triplets?

> I was using vcpkg in manifest mode and with the vcpkg toolchain file injected.

Explain how you "injected" the toolchain.

**Comment by ned14** [comment]
> You seem to use vcpkg in a wrong way.

I would doubt it.

> Why not use one of the existing static triplets?

Our triplet does custom config. Also, other projects may use the same packages as our projects, but configured differently, and we need the vcpkg binary cache to not collide.

> Explain how you "injected" the toolchain.

```
if(NOT CMAKE_TOOLCHAIN_FILE STREQUAL "${VCPKG_ROOT_DIR}/scripts/buildsystems/vcpkg.cmake")
  # This must be forced so dependency builds pick it up
  if(CMAKE_TOOLCHAIN_FILE)
    set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${CMAKE_TOOLCHAIN_FILE}" CACHE INTERNAL "vcpkg chained toolchain file")
  endif()
  set(CMAKE_TOOLCHAIN_FILE "${VCPKG_ROOT_DIR}/scripts/buildsystems/vcpkg.cmake" CACHE INTERNAL "vcpkg toolchain file" FORCE)
endif()
```
External build sets its own toolchain file to suit itself, and changing that setup is not possible.

**Comment by dg0yt** [comment]
> > You seem to use vcpkg in a wrong way.
> 
> I would doubt it.

Hm. Then everyone else must be doing it wrong, including CI.

> > Why not use one of the existing static triplets?
> 
> Our triplet does custom config. Also, other projects may use the same packages as our projects, but configured differently, and we need the vcpkg binary cache to not collide.

That is a fair reason. 
(But so it doesn't matter if and why static isn't the default on Windows.)

> > Explain how you "injected" the toolchain.
> 
> ```
> if(NOT CMAKE_TOOLCHAIN_FILE STREQUAL "${VCPKG_ROOT_DIR}/scripts/buildsystems/vcpkg.cmake")
>   # This must be forced so dependency builds pick it up
>   if(CMAKE_TOOLCHAIN_FILE)
>     set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${CMAKE_TOOLCHAIN_FILE}" CACHE INTERNAL "vcpkg chained toolchain file")
>   endif()
>   set(CMAKE_TOOLCHAIN_FILE "${VCPKG_ROOT_DIR}/scripts/buildsystems/vcpkg.cmake" CACHE INTERNAL "vcpkg toolchain file" FORCE)
> endif()
> ```
> External build sets its own toolchain file to suit itself, and changing that setup is not possible.

This fragment looks good, for the integration based on modifying `CMakeLists.txt`. Please check:
- Is this code located before the first  `project` command in your CMake project?
- Was it there already before the first CMake configuration run when you faced the problem? (Did you try with a fresh build directory?)
- Is there any other `set(CMAKE_TOOLCHAIN_FILE ...)` without `CACHE`?

**Comment by ned14** [comment]
> This fragment looks good, for the integration based on modifying CMakeLists.txt. Please check:
> 
> Is this code located before the first project command in your CMake project?

Yes

> Was it there already before the first CMake configuration run when you faced the problem? (Did you try with a fresh build directory?)

The CI always builds from fresh. And yes everything done from fresh locally too.

> Is there any other set(CMAKE_TOOLCHAIN_FILE ...) without CACHE?

No

I don't think the problem is the cmake variables set up by the vcpkg toolchain file. As I mentioned above, CONFIG based `find_package()` works fine. The problem is when `find_package()` falls back to `FindXXX.cmake`, those cmake Find scripts can be a touch idiosyncratic. To be specific, if they see multiple choices, they randomly pick one. Those are outside the control of vcpkg, and that's what is very broken here.

**Comment by dg0yt** [comment]
> The problem is when find_package() falls back to FindXXX.cmake, those cmake Find scripts can be a touch idiosyncratic. To be specific, if they see multiple choices, they randomly pick one. Those are outside the control of vcpkg, and that's what is very broken here.

There is control, via search path variables and via wrappers. The question is why it doesn't work in your setup.

It would be best to create a minimal reproducing example, e.g. for `ZLIB`, and run with `CMAKE_FIND_DEBUG_MODE` set to 1.

**Comment by dg0yt** [comment]
> It works I believe within design specifications, which is to say the `FindXXX.cmake` script will find you _some_ installation of the library in question. Which could be vcpkg's, could be the system's, could be any other copy cmake's search paths allow.
> ...
> * For https://cmake.org/cmake/help/latest/module/FindZLIB.html there is no override variable :(
> * For https://cmake.org/cmake/help/latest/module/FindOpenSSL.html there is no override variable :(

Proper find modules use `find_path` and `find_library` which a) use a defined search procedure and b) are bound to cache variables which can be set in advance - that's why they are now called "input variables" in CMake.

The vcpkg cmake toolchain configures the search paths, and it hooks into `find_package` to use wrappers e.g. for zlib and openssl which set input variables. This works for ports in vcpkg, for my user projects, and for other users, too.

But the user or project is free to override/break these features by modifying CMake variables. To find out what causes the problem in your setup, it would be necessary to get facts about your setup, not speculation about what could be impossible. Again, it would be best to create a minimal reproducing example, e.g. for ZLIB, and set CMAKE_FIND_DEBUG_MODE to 1 before `find_package(ZLIB)`.

**Comment by ned14** [comment]
> But the user or project is free to override/break these features by modifying CMake variables. To find out what causes the problem in your setup, it would be necessary to get facts about your setup, not speculation about what could be impossible. Again, it would be best to create a minimal reproducing example, e.g. for ZLIB, and set CMAKE_FIND_DEBUG_MODE to 1 before find_package(ZLIB).

I'm getting a feeling this will be a wontfix. My hacky solution fixed the problem my side of things, all CI runners are now working and the production binary images Docker images are also now working. So I've moved onto other work. I was only upcalling this fix out of courtesy.

Thanks for your help.

**Comment by dg0yt** [comment]
> I'm getting a feeling this will be a wontfix. 

Why? I really tried to make clear that this is very much a *requires-repro*. 

> I was only upcalling this fix out of courtesy.

Fix what? "Non-CONFIG find_package() is very broken in vcpkg" is a broad claim, given that the only evidence is your local setup. 

Again, it would be best to create a minimal reproducing example, e.g. for ZLIB, and set CMAKE_FIND_DEBUG_MODE to 1 before find_package(ZLIB). This should be much less work than inventing the proposed vcpkg_find_package. (You will find that the proposed vcpkg_find_package has limitations for static linkage, for the CONFIG keyword, and maybe more. That's why there are wrappers in vcpkg.)

**Comment by JackBoosY** [comment]
Sorry for late, I'm quite busy on other task yesterday, and I will try to understand this issue later.

**Comment by JackBoosY** [comment]
1. If you should use a triplet which has different configuration with the official triplet, you should write a triplet file and use feature overlay-triplet. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/examples/overlay-triplets-linux-dynamic.md).
2. If some of the ports needs to be built as static / dynamic when using dynamic / static triplet. you can add the following code to your custom triplet file:
```cmake
if (${PORT} STREQUAL "PORT_NAME_WHICH_IS_DIFFERENT")
    set(VCPKG_LIBRARY_LINKAGE "dynamic") # or "static"
endif()
```
3.  If the port you use can export the cmake configuration file, vcpkg will only **prepend** the _VCPKG_ROOT/installed/TRIPLET/share/PORT_NAME_ path to the default path where cmake finds `*Config.cmake` or `Find*.cmake` by default, and vcpkg **won't** delete the system or your settings.
Also, I assume you should use the find method provided by vcpkg because you need to use the port provided by vcpkg.
And if you need to use the system libs, you should se `-DVCPKG_PREFER_SYSTEM_LIBS=ON` in your custom project cmake configuration. See https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake#L51.
4. You should not modify the vcpkg configuration to implement the features you need, because vcpkg should already provide the corresponding features.

**Comment by ned14** [comment]
> Fix what? "Non-CONFIG find_package() is very broken in vcpkg" is a broad claim, given that the only evidence is your local setup.

The Docker images used are the public ones. We build for a wide variety of platforms. We have some older images in there e.g. centos 6, where if the system OpenSSL is found instead of the vcpkg one, link errors result. This is how we discovered this issue in fact, because on the newer images the system OpenSSL is some 1.1 variant and that's ABI compatible with the vcpkg OpenSSL. Upon deeper investigation, we found not only is OpenSSL afflicted, the wrong zlib was sometimes being chosen as well. I could see no pattern across a range of centos and ubuntu images, whether the vcpkg or the system installed library was chosen appears to not be predictable.

This isn't of course actually a surprise. There is good reason why Kitware replaced the Module find package mechanism with the Config find package mechanism. The latter is reliable and much less prone to unpleasant surprise. If you've ever had the unpleasantness of trying to use two incompatible copies of OpenSSL in the same project you'll know exactly what I mean here.

> Again, it would be best to create a minimal reproducing example, e.g. for ZLIB, and set CMAKE_FIND_DEBUG_MODE to 1 before find_package(ZLIB). This should be much less work than inventing the proposed vcpkg_find_package. (You will find that the proposed vcpkg_find_package has limitations for static linkage, for the CONFIG keyword, and maybe more. That's why there are wrappers in vcpkg.)

I appreciate that it is frustrating for you, but I don't have the time to do any further work on this. Multiple spinning plates etc. I solved the problem locally, the teams at work blocked on this are unblocked and happy, I'm now working on unblocking other things blocking other teams. I can invest no further time into this.

> Sorry for late, I'm quite busy on other task yesterday, and I will try to understand this issue later.

Be aware today is actually a day off work for me, and I will be departing for Dublin shortly. Back on Monday.

In terms of actually fixing the issue, the ideal is vcpkg requiring ports implement cmake config and nothing else is permitted. This is what cmake hunter does for the most part, and it works well. That's a hefty initial burden to land onto ports maintainers though i.e. you'd probably have to go assign one of your very few full time devs to help out with this, and I totally appreciate that this won't be high priority relative to other high priority work items. Still, I reported the issue here out of courtesy, after all we all know each other and will be seeing each other again in the near future.

**Comment by ned14** [comment]
> If you should use a triplet which has different configuration with the official triplet, you should write a triplet file and use feature overlay-triplet. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/examples/overlay-triplets-linux-dynamic.md).


We are using manifest mode, so we never call `vcpkg install`. How does one specify overlay triplet in this use case?

> If some of the ports needs to be built as static / dynamic when using dynamic / static triplet. you can add the following code to your custom triplet file:
> if (${PORT} STREQUAL "PORT_NAME_WHICH_IS_DIFFERENT")
>     set(VCPKG_LIBRARY_LINKAGE "dynamic") # or "static"
> endif()

We never want shared libraries in our use case, so static linkage is always forced on.

> Also, I assume you should use the find method provided by vcpkg because you need to use the port provided by vcpkg.

We do if you examine the code I posted above. It does indirect correctly via vcpkg's find package shim.

> And if you need to use the system libs, you should set -DVCPKG_PREFER_SYSTEM_LIBS=ON in your custom project cmake configuration. See https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake#L51.

We want to never, ever, use the system's libraries instead of a library in the vcpkg manifest.

**Comment by dg0yt** [comment]
> We are using manifest mode, so we never call `vcpkg install`. How does one specify overlay triplet in this use case?

In manifest mode with CMake, you can use CMake variables:
- `VCPKG_TARGET_TRIPLET` : set the desired triplet (e.g. `x64-window-static`)
- `VCPKG_OVERLAY_TRIPLETS`: a list of paths
- `VCPKG_INSTALL_OPTIONS`: additional options for vcpkg (e.g. `--debug`)

Reference: https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#cmake-integration

**Comment by JackBoosY** [comment]
@ned14 Yes, we prefer to use manifest mode instead of the classical mode.

> We want to never, ever, use the system's libraries instead of a library in the vcpkg manifest.

The process would be:
1. Use the cmake configuration file provided by the port.
2. Use the cmake configuration/module file you provided.
3. Use the cmake module file `Find*.cmake` provided by cmake.

I don't know what's wrong here, so can you provide a minum repro step or example?

Thanks.

**Comment by JackBoosY** [comment]
We don't have enough information to solve this issue, unfortunately; if you have any more information to help us solve this, please reopen!
